mlir/lib/Conversion/AIRRtToIpuPass.cpp

```diff
@@ -344,6 +344,34 @@ static LogicalResult CastFunctionArgs(func::FuncOp funcOp,
     rewriter.setInsertionPointToStart(&entry);
     auto cast = rewriter.create<UnrealizedConversionCastOp>(
         rewriter.getUnknownLoc(), memrefTy, entry.getArgument(i));
+    // With memref shape collapsed to 1d, the multi-dimensional offset also
+    // needs to be collapsed.
+    SmallVector<Operation *> users;
+    for (auto user : entry.getArgument(i + 1).getUsers()) {
+      if (auto cast = dyn_cast<UnrealizedConversionCastOp>(user)) {
+        assert(cast.getNumResults() == 1);
+        for (auto cast_user : cast.getResult(0).getUsers())
+          users.push_back(cast_user);
+      } else
+        users.push_back(user);
+    }
+    for (Operation *user : users) {
+      if (auto dmaUser = dyn_cast<AIEX::IpuDmaMemcpyNdOp>(user)) {
+        int oneDOffset = *getConstantIntValue(dmaUser.getMixedOffsets().back());
+        for (int i = dmaUser.getMixedOffsets().size() - 2; i >= 0; i--)
+          oneDOffset += *getConstantIntValue(dmaUser.getMixedOffsets()[i]) *
+                        *getConstantIntValue(dmaUser.getMixedStrides()[i]);
+        rewriter.setInsertionPoint(dmaUser);
+        const std::vector<int64_t> newStaticOffsets = {0, 0, 0, oneDOffset};
+        rewriter.create<AIEX::IpuDmaMemcpyNdOp>(
+            rewriter.getUnknownLoc(), dmaUser.getX(), dmaUser.getY(),
+            dmaUser.getMemref(), SmallVector<Value>{}, dmaUser.getSizes(),
+            dmaUser.getStrides(), ArrayRef(newStaticOffsets),
+            dmaUser.getStaticSizes(), dmaUser.getStaticStrides(),
+            dmaUser.getMetadata(), dmaUser.getId());
+        dmaUser->erase();
```

Comment:
This pattern already modifies the IR directly, but should use `rewriter.eraseOp` here instead of `erase`

Reply:
Done.